reactive/kotlinx-coroutines-rx2/src/RxScheduler.kt

```diff
@@ -4,16 +4,118 @@
+    }
+
+/**
+ * Converts an instance of [CoroutineDispatcher] to an implementation of [Scheduler].
+ */
+public fun CoroutineDispatcher.asScheduler(): Scheduler =
+    if (this is SchedulerCoroutineDispatcher) {
+        scheduler
+    } else {
+        DispatcherScheduler(this)
+    }
+
+private class DispatcherScheduler(internal val dispatcher: CoroutineDispatcher) : Scheduler() {
+
+    val parentJob = SupervisorJob()
+    private val parentScope = CoroutineScope(parentJob)
+
+    override fun scheduleDirect(run: Runnable): Disposable {
+        val decoratedRun = RxJavaPlugins.onSchedule(run)
+        val worker = createWorker() as DispatcherWorker
+        return worker.apply {
+            worker.schedule(decoratedRun)
+        }
+    }
+
+    override fun scheduleDirect(run: Runnable, delay: Long, unit: TimeUnit): Disposable {
+        val decoratedRun = RxJavaPlugins.onSchedule(run)
+        val worker = createWorker() as DispatcherWorker
+        return worker.apply {
+            schedule(decoratedRun, delay, unit)
+        }
+    }
+
+    override fun createWorker(): Worker =
+        DispatcherWorker(dispatcher, parentJob)
+
+    override fun shutdown() {
+        parentScope.cancel()
+    }
+
+    private class DispatcherWorker(private val dispatcher: CoroutineDispatcher, parentJob: Job) : Worker() {
+
+        private val workerScope = CoroutineScope(SupervisorJob(parentJob) + dispatcher)
+        private val blockChannel = Channel<Runnable>()
+        private var queueProcessingJob: Job? = null
+
+        override fun isDisposed(): Boolean = !workerScope.isActive
+
+        override fun schedule(block: Runnable): Disposable {
+            startProcessingQueue()
+            if (workerScope.isActive) {
+                workerScope.launch(dispatcher) {
+                    blockChannel.send(block)
+                }
+                return this
+            }
+
+            return Disposables.disposed()
+        }
+
+        override fun schedule(block: Runnable, delay: Long, unit: TimeUnit): Disposable {
+            startProcessingQueue()
+            if (delay <= 0) {
+                return schedule(block)
+            }
+            if (workerScope.isActive) {
+                workerScope.launch {
+                    delay(unit.toMillis(delay))
+                    block.run()
+                }
+                return this
+            }
+            return Disposables.disposed()
+        }
+
+        override fun dispose() {
+            workerScope.cancel()
+            blockChannel.close()
+        }
+
+        private fun startProcessingQueue() {
+            if (queueProcessingJob == null) {
+                queueProcessingJob = workerScope.launch {
+                    while (true) {
+                        if (isActive && !blockChannel.isEmpty) {
+                            val block = blockChannel.receive()
+                            block.run()
+                        }
+                        yield()
```

Comment:
Won't this loop effectively busy-wait on the channel? It seems like it would be simpler and more efficient to just receive on the channel and let the coroutine suspend until another runnable is available.